src/mlpack/core/util/sfinae_utility.hpp

```diff
@@ -182,6 +182,45 @@ struct NAME                                                                    \
+ * the passed template parameter, has a member function with the specified name.
+ * The check does not care about the signature or the function parameters
+ *
+ * @param FUNC the name of the function, whose existence is to be detected
+ * @param NAME the name of the structure that will be generated
+ *
+ * Use this like: NAME<ClassName>::value to check for the existence of the function
+ * in the given class name.
+ * This can also be used in conjunction with std::enale_if.
+ */
+#define HAS_ANY_METHOD_FORM( FUNC, NAME )                                                     \
+template < typename T >                                                                       \
+struct has_member_function                                                                    \
+{                                                                                             \
+  template < typename Q = T >                                                                 \
+  static typename                                                                             \
+  std::enable_if< std::is_member_function_pointer< decltype( &Q::FUNC ) >::value, int >::type \
+  f( int t )                                                                                  \
+  {                                                                                           \
+    return 1;                                                                                 \
+  }                                                                                           \
+                                                                                              \
+  template < typename Q = T >                                                                 \
+  static char f( char t )                                                                     \
+  {                                                                                           \
+    return 0;                                                                                 \
+  }                                                                                           \
+                                                                                              \
+  static const bool value = sizeof( f< T >( 0 ) ) != sizeof( char );                          \
+};                                                                                            \
+                                                                                              \
+template < typename T >                                                                       \
+struct NAME                                                                                   \
+{                                                                                             \
+  static const bool value = has_member_function< T >::value;                                  \
```

Comment:
Do you think you could simplify this a bit by removing the `has_member_function` struct entirely?  Maybe this works:

```
static const bool value = std::is_member_function_pointer<decltype(&T::FUNC)>::value;
```